main/init/maven/src/mill/main/maven/BuildGen.scala

```diff
@@ -0,0 +1,316 @@
+ *  - build profiles
+ */
+@mill.api.internal
+object BuildGen {
+
+  def main(args: Array[String]): Unit = {
+    val cfg = ParserForClass[BuildGenConfig].constructOrExit(args.toSeq)
+    run(cfg)
+  }
+
+  private type MavenNode = Node[Model]
+
+  private def run(cfg: BuildGenConfig): Unit = {
+    val workspace = os.pwd
+
+    println("converting Maven build")
+    val modeler = Modeler(cfg)
+    val input = Tree.from(Seq.empty[String]) { dirs =>
+      val model = modeler(workspace / dirs)
+      (Node(dirs, model), model.getModules.iterator().asScala.map(dirs :+ _))
+    }
+
+    val output = convert(input, cfg)
+    writeBuildObject(if (cfg.merge.value) compactBuildTree(output) else output)
+
+    println("converted Maven build to Mill")
+  }
+
+  private def convert(input: Tree[MavenNode], cfg: BuildGenConfig): Tree[Node[BuildObject]] = {
+    val packages = // for resolving moduleDeps
+      buildPackages(input)(model => (model.getGroupId, model.getArtifactId, model.getVersion))
+
+    val moduleSupertypes = Seq("PublishModule", "MavenModule")
+
+    val (
+      baseJavacOptions,
+      baseRepos,
+      baseNoPom,
+      basePublishVersion,
+      basePublishProperties,
+      baseModuleTypedef
+    ) = cfg.shared.baseModule match {
+      case Some(baseModule) =>
+        val model = input.node.value
+        val javacOptions = Plugins.MavenCompilerPlugin.javacOptions(model)
+        val pomSettings = extractPomSettings(model)
+        val publishVersion = model.getVersion
+        val publishProperties = getPublishProperties(model, cfg)
+
+        val typedef = IrTrait(
+          cfg.shared.jvmId,
+          baseModule,
+          moduleSupertypes,
+          javacOptions,
+          pomSettings,
+          publishVersion,
+          publishProperties
+        )
+
+        (javacOptions, Seq.empty, false, publishVersion, publishProperties, typedef)
+      case None =>
+        (Seq.empty, Seq.empty, true, "", Seq.empty, null)
+    }
+
+    input.map { case build @ Node(dirs, model) =>
+      val artifactId = model.getArtifactId
+      println(s"converting module $artifactId")
+
+      val millSourcePath = os.Path(model.getProjectDirectory)
+      val packaging = model.getPackaging
+
+      val isNested = dirs.nonEmpty
+      val hasTest = os.exists(millSourcePath / "src/test")
+
+      val supertypes =
+        Seq("RootModule") ++
+          cfg.shared.baseModule.fold(moduleSupertypes)(Seq(_))
+
+      val scopedDeps = extractScopedDeps(model, packages, cfg)
+
+      val options = Plugins.MavenCompilerPlugin.javacOptions(model)
+
+      def processResources(input: java.util.List[org.apache.maven.model.Resource]) = input
+        .asScala
+        .map(_.getDirectory)
+        .map(os.Path(_).subRelativeTo(millSourcePath))
+        .filterNot(_ == mavenTestResourceDir)
+        .toSeq
+
+      val inner = IrBuild(
+        scopedDeps,
+        cfg.shared.testModule,
+        hasTest,
+        dirs,
+        repos = Nil,
```

Comment:
@ajaychandran I noticed that the Maven generator doesn't fill in the `repos` config like the gradle generator does. Is that intentional?

Reply:
I tried adding this but `Model.getRepositories` is always empty. Added a comment documenting this quirk.